### PR TITLE
Use $ua->local_address to fill in REMOTE_ADDR/REMOTE_HOST

### PR DESCRIPTION
**Pull request by wolfsage**

### CODE REVIEW

File: lib/LWP/Protocol/PSGI.pm [hunk 1]
```diff
@@ -67,12 +68,35 @@ sub request {
...
+    if ($addr !~ /[a-z]/) {
+        # Just an IP
+        $env->{REMOTE_ADDR} = $addr;
+
+        return;
+    }
+
+    $env->{REMOTE_HOST} = $addr;
+
+    my $ip;
```
Comment (miyagawa) [inline lib/LWP/Protocol/PSGI.pm]: Unused?

File: lib/LWP/Protocol/PSGI.pm [hunk 2]
```diff
@@ -67,12 +68,35 @@ sub request {
...
+        $self->fill_in_host_addr($env);
         res_from_psgi $app->app->($env);
     } else {
         $orig{$self->{scheme}}->new($self->{scheme}, $self->{ua})->request($request);
     }
 }
 
+sub fill_in_host_addr {
+    my ($self, $env) = @_;
+    my $addr = $self->{ua}{local_address} || return;
```
Comment (miyagawa) [inline lib/LWP/Protocol/PSGI.pm]: nit: `or`

File: lib/LWP/Protocol/PSGI.pm (hunk 1 above)
Comment (wolfsage) [inline lib/LWP/Protocol/PSGI.pm]: Ah yeah, sorry

**Comment by wolfsage** [comment]
I do wonder if just allowing an %ENV override for these two things would be better?

**Comment by wolfsage** [comment]
Alternatively @rjbs thinks there's something we could do with middleware to give us control over host/addr, so maybe hold off on this to see what that looks like.

Sorry for the noise!

**Comment by miyagawa** [comment]
I do not what `local_address` is for - can you tell me the use case?

**Comment by wolfsage** [comment]
Sure, right now LWP::Protocol::PSGI just blindly fills in $env->{REMOTE_ADDR} as '127.0.0.1' and $env->{REMOTE_HOST} as 'localhost' (this is done by req_to_psgi).

This makes it hard to test things in your app mimicking different client IP addresses and hostnames (mostly different IPs is what I care about here).

**Comment by wolfsage** [comment]
Alright, so, we've been able to solve this with basically a middleware that wraps our app in our tests; so perhaps this change isn't necessary. I'd say take it (or something like it!) if you want/think it may be useful but otherwise I am all set.

Thanks, and apologies again.

Cheers

**Comment by rjbs** [comment]
Wrapper in question:

```
package RemoteMunger {
  use Moose;
  use experimental qw(signatures postderef);
  use namespace::autoclean;

  has http_remote => (
    is        => 'rw',
    clearer   => 'clear_http_remote',
    predicate => 'has_http_remote',
  );

  sub wrap ($self, $app) {
    return sub ($env) {
      if ($self->has_http_remote) {
        $env->{REMOTE_ADDR} = $self->http_remote->{addr};
        $env->{REMOTE_HOST} = $self->http_remote->{host};
      }
      $app->($env);
    }
  }
}
```

**Pull request closed without merging**